### PR TITLE
Limit task run cache key size

### PR DESCRIPTION
**Pull request by zangell44**
This PR adds a limit on the size of task run cache keys.

2,000 seemed like a reasonable number. Upwards of 8,000 is where Postgres will not be able to write the row because it can't fit that much data in the `cache_key` index.

See writeup here https://github.com/PrefectHQ/nebula/issues/2343.

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

File: src/prefect/orion/schemas/core.py [hunk 1]
```diff
@@ -364,6 +364,7 @@ class TaskRun(ORMBaseModel):
     )
     cache_key: Optional[str] = Field(
         default=None,
+        max_length=2_000,
```
Comment (madkinsz) [inline src/prefect/orion/schemas/core.py:367]: Can you move this to a setting so OSS servers can change it if they want and add it to the tasks concept documentation?

**Comment by madkinsz** [comment]
@serinamarie would you mind finishing this one off?

**Comment by ahuang11** [comment]
Tests seem to be failing because of a new httpx version; https://github.com/PrefectHQ/prefect/pull/8004#issuecomment-1369457371

**Review by tpdorsey — APPROVED** [review]
Docs changes look good. Ty

File: src/prefect/orion/schemas/core.py [hunk 2]
```diff
@@ -436,6 +436,14 @@ class TaskRun(ORMBaseModel):
         default=None, description="The current task run state."
     )
 
+    @validator("cache_key", pre=True)
+    def validate_cache_key_length(cls, cache_key):
+        if len(cache_key) > PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH.value():
+            raise ValueError(
+                f"Invalid cache_key length. Task cache key length can be modified using the PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH setting."
```
Comment (madkinsz) [inline src/prefect/orion/schemas/core.py]: Can we clarify that this is a server-side setting?

File: src/prefect/orion/schemas/core.py [hunk 3]
```diff
@@ -389,7 +390,6 @@ class TaskRun(ORMBaseModel):
         description="A dynamic key used to differentiate between multiple runs of the same task within the same flow run.",
     )
     cache_key: Optional[str] = Field(
-        default=None,
```
Comment (madkinsz) [inline src/prefect/orion/schemas/core.py:392]: Why are we removing the default value? This seems incorrect.

File: src/prefect/orion/schemas/core.py [hunk 4]
```diff
@@ -467,6 +467,16 @@ class TaskRun(ORMBaseModel):
     def set_name(cls, name):
         return name or generate_slug(2)
 
+    @validator("cache_key")
+    def validate_cache_key_length(cls, cache_key):
+        if cache_key:
+            if len(cache_key) > PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH.value():
```
Comment (madkinsz) [inline src/prefect/orion/schemas/core.py:473]: let's just use an `and` hre?

File: src/prefect/orion/schemas/core.py [hunk 5]
```diff
@@ -467,6 +467,16 @@ class TaskRun(ORMBaseModel):
...
+                raise ValueError(
+                    "Invalid cache_key length. The task cache key length can be modified using the PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH setting, "
+                    "which means that you can change the maximum length of the cache_key by modifying the value of this setting on the server."
```
Comment (madkinsz) [inline src/prefect/orion/schemas/core.py:476]: I think this is a little too verbose. I'd say

"Cache key exceeded maximum allowed length of {PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH} characters."

File: src/prefect/settings.py [hunk 6]
```diff
@@ -876,6 +876,9 @@ def default_cloud_ui_url(settings, value):
 until a resume attempt.
 """
 
+PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH = Setting(int, default=2000)
+"""Limit size of task run cache keys."""
```
Comment (madkinsz) [inline src/prefect/settings.py:880]:
```suggestion
"""
The maximum number of characters allowed for a task run cache key.

This setting cannot be changed client-side, it must be set on the server.
"""
```

File: src/prefect/orion/schemas/core.py (hunk 3 above)
Comment (serinamarie) [reply src/prefect/orion/schemas/core.py:392]: It is incorrect.

File: src/prefect/orion/schemas/core.py (hunk 2 above)
Comment (serinamarie) [reply src/prefect/orion/schemas/core.py]: Shortened to "Cache key exceeded maximum allowed length of {PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH} characters."

File: src/prefect/orion/schemas/core.py [hunk 7]
```diff
@@ -467,6 +468,17 @@ class TaskRun(ORMBaseModel):
...
         return name or generate_slug(2)
 
+    @validator("cache_key")
+    def validate_cache_key_length(cls, cache_key):
+        if (
+            cache_key
+            and len(cache_key) > PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH.value()
+        ):
+            raise ValueError(
+                "Cache key exceeded maximum allowed length of {PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH} characters."
```
Comment (madkinsz) [inline src/prefect/orion/schemas/core.py:478]: Probably going to exceed line length limits

```suggestion
                f"Cache key exceeded maximum allowed length of {PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH.value()} characters."
```

File: tests/orion/schemas/test_core.py [hunk 8]
```diff
@@ -171,6 +172,85 @@ async def test_flow_run_policy_populates_new_properties_from_deprecated(self):
         assert new_policy.retry_delay == 2
 
 
+class TestTaskRun:
+    def test_task_run_cache_key_greater_than_user_configured_max_length(self):
+        with temporary_settings({PREFECT_ORION_TASK_CACHE_KEY_MAX_LENGTH: 5}):
+
+            cache_key_invalid_length = "X" * 6
+            with pytest.raises(pydantic.ValidationError):
```
Comment (madkinsz) [inline tests/orion/schemas/test_core.py:180]: Let's include a `match=` clause here to check the error message is correct

**Pull request merged**